android/canonical/app/src/androidTest/java/com/google/samples/quickstart/canonical/DatabaseInteractionTest.kt

```diff
@@ -0,0 +1,83 @@
+package com.google.samples.quickstart.canonical
+
+import androidx.test.espresso.Espresso.onData
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.assertion.ViewAssertions
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.RootMatchers.isDialog
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.rule.ActivityTestRule
+import androidx.test.uiautomator.UiDevice
+import androidx.test.uiautomator.UiSelector
+import org.hamcrest.Matchers
+import org.junit.After
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class DatabaseInteractionTest {
+    @get:Rule
+    var activityRule: ActivityTestRule<MainActivity>
+            = ActivityTestRule(MainActivity::class.java)
+
+    lateinit var device : UiDevice
+
+    @Before
+    fun setup() {
+        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+        onView(ViewMatchers.withId(R.id.sign_in_button))
+            .perform(click())
+        val googleSignInDialog = device.findObject(UiSelector().text("yl4324@columbia.edu"))
+        googleSignInDialog.clickAndWaitForNewWindow()
+        // Make sure that:
+        // 1. Your google account must have signed out before test.
+        // 2. You should have at lest one google account for your device,
+        // which means, when you click sign in button, you have at least
+        // one account to choose
+    }
+
+    @Test
+    fun submitRecord() {
+        onView(ViewMatchers.withId(R.id.start_pause_btn))
+            .perform(click())
+        Thread.sleep(5000)
```

Comment:
Is Thread.sleep the preferred method or is there a method that will wait for a view to appear? I'm asking because I'm not sure, but it seems like there should be. That way tests could run as fast as possible without arbitrary delays.

Reply:
I usually did this through idling resources.

Reply:
In this case, the Thread.sleep(5000)  is not for waiting. it is used for recording 5s during the test about stopwatch. 
In fact, most of the Thread.sleep() is used for me haha. I need some time to watch the result of my test. Otherwise, it is too fast to see.

https://developer.android.com/training/testing/ui-testing/espresso-testing

`Espresso detects when the main thread is idle, so it is able to run your test commands at the appropriate time, improving the reliability of your tests. This capability also relieves you from having to add any timing workarounds, such as Thread.sleep() in your test code.`